src/main/target/STM32_UNIFIED/STM32F7X2.mk

```diff
@@ -0,0 +1,3 @@
+# Override the OPTIMISE_SPEED compiler setting to save flash space on this 512KB target.
+# Performance is only slightly affected but a very large amount of flash is saved.
+OPTIMISE_SPEED = -O2
```

Comment:
Interesting that there is so little difference between `-Ofast` and `-O2` on the architecture of the F7 - on F3 this made for around 25% performance penalty.
But then a main difference between the optimisations is that `-O3` (and better) do aggressive loop unrolling (that also yield the size penalty) to remove pipeline flushes, and it's possible that the F7's core does better prediction than the F3's.
Out of interest, did you test this with F4, and how did it perform?
Also, should we be moving this out of target specific optimisations, and into per-MCU optimisations under `/make`?

Reply:
I haven't yet tested with F4. F405 clearly doesn't need the space and F411 can't afford any performance loss so didn't think it worthwhile. I don't know Makefile syntax terribly well so I'm open to other ways of doing this. The reason I did it this way was to give the developer control over the optimization rather than "bake" it into the Makefile. For example it could be optionally set in a legacy target as well. Also was trying to avoid generalisms like "all F7" or "F7 with 512KB or less", and so on.

Reply:
I tested on a F411 and it saves ~58KB of flash. But the performance hit is more substantial and is getting borderline for loop time stability on 8/4/DSHOT300 w/bidirectional. This was tested on a board with SPI rx as that's common for this MCU and adds a fair bit of load. So while it _can_ run with optimizations reduced I'm not sure it will be stable once the user starts enabling a bunch of features like LEDSTRIP and many OSD elements. So I'd only do this out of desperation for flash space on a F411.

Before:
```
# tasks
Task list             rate/hz  max/us  avg/us maxload avgload  total/ms
00 - (         SYSTEM)     10       1       1    0.5%    0.5%         0
01 - (         SYSTEM)    989       2       1    0.6%    0.5%        23
02 - (           GYRO)   8000      17       9   14.1%    7.7%      1689
03 - (         FILTER)   4000      36      28   14.9%   11.7%      2744
04 - (            PID)   4000      91      77   36.9%   31.3%      7563
05 - (            ACC)    987      12       9    1.6%    1.3%       219
06 - (       ATTITUDE)    100      12      10    0.6%    0.6%        22
07 - (             RX)     33      62      59    0.7%    0.6%        48
08 - (         SERIAL)    100    1461       2   15.1%    0.5%      1053
09 - (       DISPATCH)    994       2       0    0.6%    0.0%        18
10 - (BATTERY_VOLTAGE)     50       3       2    0.5%    0.5%         2
11 - (BATTERY_CURRENT)     50       1       1    0.5%    0.5%         1
12 - ( BATTERY_ALERTS)      5       2       2    0.5%    0.5%         0
13 - (         BEEPER)    100       2       2    0.5%    0.5%         3
23 - (            OSD)     60      44      23    0.7%    0.6%        34
25 - (            CMS)     20       4       2    0.5%    0.5%         0
26 - (        VTXCTRL)      5       1       1    0.5%    0.5%         0
27 - (        CAMCTRL)      5       1       1    0.5%    0.5%         0
29 - (    ADCINTERNAL)      2       3       1    0.5%    0.5%         0
RX Check Function                   5       4                       237
Total (excluding SERIAL)                        75.2%   58.8%
```
After:
```
# tasks
Task list             rate/hz  max/us  avg/us maxload avgload  total/ms
00 - (         SYSTEM)     10       1       1    0.5%    0.5%         0
01 - (         SYSTEM)    988       3       1    0.7%    0.5%        51
02 - (           GYRO)   8000      18      10   14.9%    8.5%      4266
03 - (         FILTER)   4000      44      36   18.1%   14.9%      8114
04 - (            PID)   4000      98      82   39.7%   33.3%     18888
05 - (            ACC)    990      18      10    2.2%    1.4%       582
06 - (       ATTITUDE)    100      12       9    0.6%    0.5%        55
07 - (             RX)     33      74      71    0.7%    0.7%       133
08 - (         SERIAL)    100   38896       2  389.4%    0.5%      1115
09 - (       DISPATCH)    989       3       0    0.7%    0.0%        41
10 - (BATTERY_VOLTAGE)     50       3       2    0.5%    0.5%         5
11 - (BATTERY_CURRENT)     50       2       1    0.5%    0.5%         2
12 - ( BATTERY_ALERTS)      5       3       2    0.5%    0.5%         0
13 - (         BEEPER)    100       3       1    0.5%    0.5%         7
23 - (            OSD)     60      44      24    0.7%    0.6%        83
25 - (            CMS)     20       3       1    0.5%    0.5%         1
26 - (        VTXCTRL)      5       1       0    0.5%    0.0%         0
27 - (        CAMCTRL)      5       1       0    0.5%    0.0%         0
29 - (    ADCINTERNAL)      1       3       2    0.5%    0.5%         0
RX Check Function                   6       4                       241
Total (excluding SERIAL)                        82.8%   63.9%
```

Reply:
Thanks for testing, and a shame it is not viable for F411. This also supports the theory that newer ARM cores benefit less from the more aggressive speed optimisations that gcc can do.

Reply:
My idea was to add the optimisation setting as a default for all STM32F7X2 builds (and only for those), like:

```
+++ b/make/mcu/STM32F7.mk
@@ -145,6 +145,7 @@ STARTUP_SRC     = startup_stm32f746xx.s
 MCU_FLASH_SIZE  := 1024
 else ifeq ($(TARGET),$(filter $(TARGET),$(F7X2RE_TARGETS)))
 DEVICE_FLAGS   += -DSTM32F722xx
+ifeq ($(OPTIMISE_SPEED),)
+OPTIMISE_SPEED  := -O2
+endif
 ifndef LD_SCRIPT
 LD_SCRIPT       = $(LINKER_DIR)/stm32_flash_f722.ld
 endif
```

This way, all builds for the STM32F7X2 (which only comes in a 512kB flash package in a form viable for us) will use this optimisation by default, leading to consistent behaviour.
Users can then still elect to override the setting in `target.mk` for custom builds.

Reply:
I'm fine with this - just means that we have a sort of "double override" logic with the `make/mcu/STM32F7.mk` overriding the default in `Makefile` which that also potentially being overridden by the target level make file. Will need to some comments to explain what's going on as it wouldn't necessarily be obvious.

Reply:
Also note this will require a slight reordering of the `Makefile` as the mcu specific make files are currently included after the section that sets the compiler optimization levels.

Reply:
And doing this reordering might seem to break the intent of current ordering based on this comment:
```
# Default Tool options - can be overridden in {mcu}.mk files.
```
Which seems to imply that changing the debug and optimization settings can/should be done in the mcu make files. The problem is that the ordering the way it is results in `OPTIMISE_SPEED` being set in the main Makefile so the `ifeq` logic always fails.